syntax = "proto3";

package idpass;
option java_package = "org.idpass.api.proto";
option java_multiple_files = true;

message Date {
    // Year of date. Must be from 1 to 9999, or 0 if specifying a date without
    // a year.
    int32 year = 1;
    // Month of year. Must be from 1 to 12, or 0 if specifying a year without a
    // month and day.
    int32 month = 2;
    // Day of month. Must be from 1 to 31 and valid for the year and month, or 0
    // if specifying a year by itself or a year and month where the day is not
    // significant.
    int32 day = 3;
}

message Pair {
    string key = 1;
    string value = 2;
}

message Dictionary {
    repeated Pair pairs = 1;
}

message CardAccess {
    enum FingerType {
        R1 = 0;
        R2 = 1;
        R3 = 2;
        R4 = 3;
        R5 = 4;
        L1 = 5;
        L2 = 6;
        L3 = 7;
        L4 = 8;
        L5 = 9;
    }

    enum EyeType {
        R = 0;
        L = 1;
    }

    FingerType finger = 1;
    bytes fingerprint = 2;
    EyeType eye = 3;
    bytes iris = 4;
    bytes face = 5;
    string pin = 6;
}

message CardDetails {
    string surName = 1;
    string givenName = 2;
    string placeOfBirth = 3;
    Date dateOfBirth = 4;
    repeated Pair extra = 5;
    int64 createdAt = 6; // Timestamp in seconds
}

message IDPassCard {
    CardAccess access = 1;
    CardDetails details = 2;
    bytes encryptionKey = 3; // ed22519 key
}

message SignedIDPassCard {
    IDPassCard card = 1;
    bytes signature = 2;
    bytes signerPublicKey = 3;
}

message PublicSignedIDPassCard {
    CardDetails details = 1;
    bytes signature = 2;
    bytes signerPublicKey = 3;
}

message IDPassCards {
    PublicSignedIDPassCard publicCard = 1;
    bytes encryptedCard = 2;
}
